include/seqan/blast/blast_base.h

```diff
@@ -0,0 +1,669 @@
+ * @headerfile <seqan/blast.h>
+ *
+ * @tag BlastProgramTag#BlastProgramTagTBlastX
+ * @brief Tag for @link BlastProgram::TBLASTX @endlink
+ * @headerfile <seqan/blast.h>
+ *
+ * @tag BlastProgramTag#BlastProgramTagUnknown
+ * @brief Tag for @link BlastProgram::UNKNOWN @endlink
+ * @headerfile <seqan/blast.h>
+ */
+
+template <BlastProgram p>
+using BlastProgramTag = std::integral_constant<BlastProgram, p>;
+
+typedef BlastProgramTag<BlastProgram::BLASTN>  BlastProgramTagBlastN;
+typedef BlastProgramTag<BlastProgram::BLASTP>  BlastProgramTagBlastP;
+typedef BlastProgramTag<BlastProgram::BLASTX>  BlastProgramTagBlastX;
+typedef BlastProgramTag<BlastProgram::TBLASTN> BlastProgramTagTBlastN;
+typedef BlastProgramTag<BlastProgram::TBLASTX> BlastProgramTagTBlastX;
+typedef BlastProgramTag<BlastProgram::UNKNOWN> BlastProgramTagUnknown;
+
+// ============================================================================
+// Functions
+// ============================================================================
+
+// ----------------------------------------------------------------------------
+// getBlastProgram()
+// ----------------------------------------------------------------------------
+
+/*
+ * Given both an enum value and a tag, return the tag's value unless it is UNKNOWN
+ * in which case the enum's value is returned. Can be used in implementations that
+ * want to cover compile-time constants and run-time settings both.
+ */
+
+// this is always evaluated at compile-time
+template <BlastProgram p>
+constexpr BlastProgram
+getBlastProgram(BlastProgram const, BlastProgramTag<p> const &)
+{
+    return p;
+}
+
+// this will likely be evaluated at run-time
+constexpr BlastProgram
+getBlastProgram(BlastProgram const p, BlastProgramTag<BlastProgram::UNKNOWN> const &)
+{
+    return p;
+}
+
+// ----------------------------------------------------------------------------
+// qHasRevComp()
+// ----------------------------------------------------------------------------
+
+/*!
+ * @fn BlastProgramTag#qHasRevComp
+ * @brief Whether the reverse complement of the <b>query</b> sequence(s) is searched with the given BlastProgram
+ * @signature constexpr bool qHasRevComp(BlastProgram const p);
+ * template <BlastProgram p>
+ * constexpr bool qHasRevComp(BlastProgramTag<p> const &);
+ * template <BlastProgram p>
+ * constexpr bool qHasRevComp(BlastProgram const _p, BlastProgramTag<p> const &);
+ *
+ * @section Remarks
+ *
+ * This function can be used to check whether the defined BlastProgam implies the existence/creation/... of a
+ * reverse complement sequence of the query sequence.
+ *
+ * If both a tag and an enum are specified, the tag is given preference, unless it is set to
+ * @link BlastProgram::UNKNOWN @endlink.
+ *
+ * @return false for @link BlastProgram::BLASTP @endlink and @link BlastProgram::TBLASTN @endlink
+ * @return true otherwise
+ * @headerfile <seqan/blast.h>
+ */
+
+constexpr bool
+qHasRevComp(BlastProgram const p)
```

Comment:
> These should have been metafunctions.

They are used both as regular and as constexpr functions, so any other solution would have meant more code, less clarity :)